File: src/main/java/dev/shreyasayyengar/menuapi/menu/StandardMenu.java
package dev.shreyasayyengar.menuapi.menu;

import com.google.common.base.Preconditions;
import dev.shreyasayyengar.menuapi.action.StandardMenuCloseAction;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Optional;

/**
 * A standard menu is a menu that is not paginated and has a fixed size. It is the most basic type of menu for
 * bukkit development. They are constructed using a builder pattern.
 */
@SuppressWarnings({"unused", "UnusedReturnValue"})
public class StandardMenu extends Menu<StandardMenu> {

    private final Inventory bukkitInventory;
    private final Map<Integer, MenuItem> items = new HashMap<>();
    private StandardMenuCloseAction closeAction;

    /**
     * Constructs a new menu with the specified title and size.
     *
     * @param title The title of the menu.
     * @param size  The number of slots in the menu.
     */
    public StandardMenu(String title, int size) {
        super(title, size);
        this.bukkitInventory = Bukkit.createInventory(null, size, title);
    }

    private StandardMenu(StandardMenu menu) {
        super(menu.title, menu.size);
        this.bukkitInventory = Bukkit.createInventory(null, this.size, title);
        this.items.putAll(menu.items);
        this.items.forEach((itemSlot, itemMenuItem) -> bukkitInventory.setItem(itemSlot, itemMenuItem.getItemStack()));
        this.closeAction = menu.closeAction;
        this.cancelClickEventsByDefault = menu.cancelClickEventsByDefault;
        this.overriddenInventoryClickAction = menu.overriddenInventoryClickAction;
    }

    public static StandardMenu copy(StandardMenu copyMenu) {
        return new StandardMenu(copyMenu);
    }

    /**
     * Adds a MenuItem to the menu at the given slot.
     *
     * @param slot The slot to add the item to.
     * @param item The MenuItem to add.
     * @return the StandardMenu instance.
     */
    public StandardMenu setItem(int slot, MenuItem item) {
        if (slot == -1) return this; // likely passed through from getNextEmptySlot which returned -1;

        // if slot already contains item, replace it:
        if (this.items.containsKey(slot)) {
            removeItem(slot);
        }

        this.items.put(slot, item);
        this.items.forEach((itemSlot, itemMenuItem) -> bukkitInventory.setItem(slot, item.getItemStack()));
        return this;
    }

    /**
     * Adds a map of items to the menu.
     *
     * @param items The map of items to add to the menu.
     * @return the StandardMenu instance.
     */
    public StandardMenu setItems(Map<Integer, MenuItem> items) {
        this.items.putAll(items);
        this.items.forEach((itemSlot, itemMenuItem) -> bukkitInventory.setItem(itemSlot, itemMenuItem.getItemStack()));
        return this;
    }

    /**
     * Removes an item from the menu at the given slot.
     *
     * @param slot The slot to remove the item from.
     * @return the StandardMenu instance.
     */
    public StandardMenu removeItem(int slot) {
        this.items.remove(slot);
        bukkitInventory.setItem(slot, null);
        return this;
    }

    /**
     * Sets the action to perform when a player closes the menu.
     *
     * @param closeAction The action to perform.
     * @return the StandardMenu instance.
     */
    public StandardMenu onClose(StandardMenuCloseAction closeAction) {
        this.closeAction = closeAction;
        return this;
    }

    /**
     * Opens the menu for the specified players. If syncMenus is true, the physical menu displayed to the client will be synchronized
     * with all players passed into the method. This means any ItemStacks added or removed from the menu will be reflected to every player passed in.
     *
     * @param syncMenus Whether to synchronize the menu with all players
     * @param players   The players to open the menu for
     */
    // TODO check functiona1lity with calling this method for different groups of players (the first menu will have to be a copyOf)
    public void open(boolean syncMenus, Player... players) {
        Preconditions.checkState(this.size != 0, "Menu size must be set or cannot be 0");

        if (syncMenus) {
            this.items.forEach((slot, item) -> bukkitInventory.setItem(slot, item.getItemStack()));
            for (Player player : players) {
                player.openInventory(bukkitInventory);
                MenuManager.getManagerInstance().getOpenMenus().put(player.getUniqueId(), this);
            }
        } else {
            for (Player player : players) {
                StandardMenu menuCopy = StandardMenu.copy(this);
                menuCopy.items.forEach((slot, item) -> menuCopy.bukkitInventory.setItem(slot, item.getItemStack()));
                player.openInventory(menuCopy.bukkitInventory);
                MenuManager.getManagerInstance().getOpenMenus().put(player.getUniqueId(), menuCopy);
            }
        }
    }

    /**
     * Opens the menu for the specified player.
     *
     * @param player The player to open the menu for
     */
    public void open(Player player) {
        open(false, player);
    }

    // --------- Internal ---------

    /**
     * {@inheritDoc}
     */
    @Override
    protected void handleClose(InventoryCloseEvent event) {
        if (this.closeAction == null) return;
        this.closeAction.performCloseAction((Player) event.getPlayer(), event.getInventory(), event);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Optional<MenuItem> getItem(int slot) {

        if (this.items.get(slot) == null || bukkitInventory.getItem(slot) == null) {
            return Optional.empty();
        }

        return Optional.of(this.items.get(slot));
    }

    /**
     * @return The next empty slot in the menu or -1 if there are no empty slots.
     */
    // TODO move this to Menu base class as an abstract method?
    public int getNextEmptySlot() {
        for (int i = 0; i < this.size; i++) {
            if (!this.items.containsKey(i)) {
                return i;
            }
        }
        return -1;
    }

    @Override
    public @NotNull Iterator<ItemStack> iterator() {
        return this.items.values().stream().map(MenuItem::getItemStack).iterator();
    }
}

File: src/main/kotlin/dev/shreyasayyengar/menuapi/menu/StandardMenu.kt
package dev.shreyasayyengar.menuapi.menu

import net.kyori.adventure.text.Component
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import java.util.*

/**
 * Represents a standard menu with a fixed size.
 *
 * @param title The title of the menu.
 * @param size The size of the menu.
 */
class StandardMenu(
    title: Component,
    size: Int,
) : Menu(title, size) {
    private var closeAction: CloseAction? = null
    private val items: MutableMap<Int, MenuItem> = mutableMapOf()
    private val bukkitInventory: Inventory = Bukkit.createInventory(null, size, title)

    companion object {
        /**
         * Copies the given [StandardMenu] instance.
         *
         * @param standardMenu The [StandardMenu] instance to copy.
         * @return An identical [StandardMenu] instance with the same properties as the given [StandardMenu].
         */
        fun copy(standardMenu: StandardMenu): StandardMenu {
            return StandardMenu(standardMenu.title, standardMenu.size).apply {
                cancelClickEventsByDefault = standardMenu.cancelClickEventsByDefault
                overriddenClickAction = standardMenu.overriddenClickAction
                handleMenuItemAction = standardMenu.handleMenuItemAction
                closeAction = standardMenu.closeAction
                items.putAll(standardMenu.items)
            }
        }
    }

    override fun handleClose(event: InventoryCloseEvent) {
        if (this.closeAction == null) return
        this.closeAction!!.performCloseAction(event.player as Player, event.inventory, event)
    }

    override fun getItem(slot: Int): Optional<MenuItem> {
        return if (this.items[slot] == null || bukkitInventory.getItem(slot) == null) Optional.empty()
        else Optional.of(this.items[slot]!!)
    }

    override fun iterator(): Iterator<ItemStack> = this.items.values.map { it.itemStack() }.iterator()

    /**
     * Adds a [MenuItem] to the menu at the specified slot.
     * If an item already exists at the specified slot, it will be replaced.
     *
     * @param slot The slot to add the item to.
     * @param item The [MenuItem] to add.
     * @return This [StandardMenu] instance.
     */
    fun addItem(slot: Int, item: MenuItem) = apply {
        require(slot in 0 until size) { "Item slot must be between 0 and ${size - 1}" }
        this.items[slot] = item
        this.items.forEach { (slot, menuItem) -> bukkitInventory.setItem(slot, menuItem.itemStack()) }
    }

    /**
     * Removes the [MenuItem] at the specified slot.
     *
     * @param slot The slot to remove the item from.
     * @return This [StandardMenu] instance.
     */
    fun removeItem(slot: Int) = apply {
        items.remove(slot)
        bukkitInventory.setItem(slot, null)
    }

    /**
     * Sets the close action for this menu.
     *
     * @param action The action to perform when the menu is closed.
     * @return This [StandardMenu] instance.
     * @see CloseAction
     */
    fun onClose(action: CloseAction) = apply { this.closeAction = action }

    /**
     * Opens this menu for the specified player(s).
     *
     * @param player The player(s) to open the menu for.
     * @see open
     */
    fun open(vararg player: Player) = open(false, *player)

    /**
     * Opens this menu for the specified player(s).
     *
     * @param syncMenus Whether to synchronize the menu for all players.
     * @param player The player(s) to open the menu for.
     */
    fun open(syncMenus: Boolean, vararg player: Player) {
        if (syncMenus || player.size == 1) {
            this.items.forEach { (slot, menuItem) -> bukkitInventory.setItem(slot, menuItem.itemStack()) }
            player.forEach {
                it.openInventory(bukkitInventory)
                MenuAPI.getInstance().openMenus[it.uniqueId] = this
            }
        } else {
            player.forEach {
                val clonedMenu = copy(this)
                clonedMenu.items.forEach { (slot, menuItem) -> clonedMenu.bukkitInventory.setItem(slot, menuItem.itemStack()) }

                it.openInventory(clonedMenu.bukkitInventory)
                MenuAPI.getInstance().openMenus[it.uniqueId] = clonedMenu
            }
        }
    }

    @FunctionalInterface
    fun interface CloseAction {
        fun performCloseAction(whoClosed: Player, inventory: Inventory, event: InventoryCloseEvent)
    }
}
